src/gluonts/transform/sampler.py

```diff
@@ -19,28 +21,29 @@
 
 class InstanceSampler:
     """
-    An InstanceSampler is called with the time series and the valid
-    index bounds a, b and should return a set of indices a <= i <= b
-    at which training instances will be generated.
-
-    The object should be called with:
+    An InstanceSampler is called with the time series ``ts``, and returns
+    a set of indices at which training instances will be generated.
 
-    Parameters
-    ----------
-    ts
-        target that should be sampled with shape (dim, seq_len)
-    a
-        first index of the target that can be sampled
-    b
-        last index of the target that can be sampled
-
-    Returns
-    -------
-    np.ndarray
-        Selected points to sample
+    The sampled indices ``i`` satisfy ``a <= i <= b``, where ``a = skip_initial``
+    and ``b = ts.shape[axis] - skip_final``.
     """
 
-    def __call__(self, ts: np.ndarray, a: int, b: int) -> np.ndarray:
+    def __init__(
+        self, *, axis: int = -1, skip_initial: int = 0, skip_final: int = 0
```

Comment:
This is very flexible and much nicer than the previous bounds approach. But perhaps we can pass additional parameters to `__call__`, such as `context_length`, `past_length`, etc.

The `skip_initial` is an important parameter here that should perhaps be set to `context_length` by default. Otherwise the network may try to predict something from nothing (I know this was the previous behavior...). Also perhaps `min_history` is a better name than `skip_initial`?

Reply:
In my mind, `__call__` should take anything that is entry-dependent, while other fixed settings (that are usually tied to some model hyperparameters) are better set in the constructor: this way we don’t force additional arguments in the API, while each concrete sampler is free to take as many configuration options as it wants.

About the default `skip_initial` value: sure, we can decide to use `context_length` as default, but I think that can be set in each estimator’s default sampler. 

I like `min_history` as name, it tells exactly what requirement the sampled indices will have with respect to the given sequence, while being detached from model-specific concepts (like `context_length`).

What about `skip_final`?

Reply:
I agree about what should be passed to call in principle, however in this case it makes it difficult to write samplers that have a clear behavior, because they need to know things like `context_length` etc, unless we want to go with writing sampler factories, using functools.partial. or setting some attributes of the sampler later. I think I ran into this with my other comment too, where you cannot instantiate a sampler without knowing `prediction_length`.

Reply:
I'm not sure there's a clear advantage in having `context_length` or other quantities passed to `__call__` rather than used at construction time. The potential drawbacks I see are the following:
* models may have multiple "time lengths" as hyperparameters, so I'm afraid that pinning down what exactly should be provided to `__call__` may not always be clear
* moreover, the sampler is used through the splitter, so in order to specify which quantities the splitter will pass on to the sampler, the interfaces of these two objects must be in sync, which may make things too implicit

In general, it's my impression that the instance splitter class should be extremely simple and general. We might not even need to have two separate ones (`InstanceSplitter` and `CanonicalInstanceSplitter`)

Reply:
Ok, yeah maybe its more about documentation and good defaults. We can just set the `min_history` to `context_length` when we construct the sampler inside all models, as this should be the default behavior. Then if you provide a sampler yourself you should know what you are doing.

Reply:
> What about `skip_final`?

Yeah, or something like `min_future`.